analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.IsNull.cs

```diff
@@ -0,0 +1,208 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2022 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.SymbolicExecution.Constraints;
+using SonarAnalyzer.UnitTest.TestFramework.SymbolicExecution;
+
+namespace SonarAnalyzer.UnitTest.SymbolicExecution.Roslyn
+{
+    public partial class RoslynSymbolicExecutionTest
+    {
+        [TestMethod]
+        public void IsNull_Coalesce_SetsObjectConstraint()
+        {
+            const string code = @"
+object nullValue = null;
+object notNullValue = new object();
+var nullToNull = nullValue ?? nullValue;
+var nullToNotNull = nullValue ?? notNullValue;
+var nullToUnknown = nullValue ?? arg;
+var notNullToNull = notNullValue ?? nullValue;
+var notNullToNotNull = notNullValue ?? notNullValue;
+var notNullToUnknown = notNullValue ?? arg;
+Tag(""NullToNull"", nullToNull);
+Tag(""NullToNotNull"", nullToNotNull);
+Tag(""NullToUnknown"", nullToUnknown);
+Tag(""NotNullToNull"", notNullToNull);
+Tag(""NotNullToNotNull"", notNullToNotNull);
+Tag(""NotNullToUnknown"", notNullToUnknown);";
+            var validator = SETestContext.CreateCS(code, ", object arg").Validator;
+            validator.ValidateContainsOperation(OperationKind.IsNull);
+            validator.ValidateTag("NullToNull", x => x.HasConstraint(ObjectConstraint.Null).Should().BeTrue());
+            validator.ValidateTag("NullToNotNull", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("NullToUnknown", x => x.Should().BeNull());
+            validator.ValidateTag("NotNullToNull", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("NotNullToNotNull", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
+            validator.ValidateTag("NotNullToUnknown", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue());
```

Comment:
Do we have a rule for this? "'{0}' is known to be not null, and {1} is unreachable. Remove the redundant {1}."

Reply:
Yes, S2589 and S2583. They will be migrated in MMF-2563

Reply:
I'm not sure if we already support this case and if we have test cases for it.